drivers/gpio/gpio_pinctrl_stm32.c

```diff
@@ -0,0 +1,685 @@
+			| GPIO_PORT_PIN12 | GPIO_PORT_PIN13 | GPIO_PORT_PIN14
+			| GPIO_PORT_PIN15;
+		pin_mask = GPIO_PORT_PIN0;
+	}
+
+	while (pin != 0) {
+		u32_t gpio_port_pin = pin & pin_mask;
+		pin &= ~pin_mask; /* remove this pin from pin(s) */
+		pin_mask = pin_mask << 1; /* prepare for next pin */
+		if (!gpio_port_pin) {
+			continue;
+		}
+		int err;
+		u16_t pinctrl_pin = cfg->pinctrl_base
+					+ gpio_port_pin_idx(gpio_port_pin);
+
+		/* request pin */
+		err = pinctrl_mux_request(data->pinctrl, pinctrl_pin,
+					  dev->config->name);
+		if (err) {
+			return err;
+		}
+
+		/* convert GPIO flags to PINCTRL pin configuration
+		 * and configure pin */
+		u32_t config = 0;
+		u16_t func;
+		err = pinctrl_config_get(data->pinctrl, pinctrl_pin, &config);
+		if (err) {
+			return err;
+		}
+		config &= ~PINCTRL_CONFIG_INPUT_MASK;
+		config &= ~PINCTRL_CONFIG_OUTPUT_MASK;
+		if ((flags & GPIO_DIR_MASK) == GPIO_DIR_OUT) {
+			config |= PINCTRL_CONFIG_INPUT_ENABLE;
+			config |= PINCTRL_CONFIG_OUTPUT_ENABLE;
+			func = PINCTRL_STM32_FUNCTION_OUTPUT;
+		} else {
+			config |= PINCTRL_CONFIG_INPUT_ENABLE;
+			config |= PINCTRL_CONFIG_OUTPUT_DISABLE;
+			func = PINCTRL_STM32_FUNCTION_INPUT;
+		}
+		if ((flags & GPIO_POL_MASK) == GPIO_POL_INV) {
+			config &= ~PINCTRL_CONFIG_OUTPUT_HIGH;
+			config |= PINCTRL_CONFIG_OUTPUT_LOW;
+		} else {
+			config &= ~PINCTRL_CONFIG_OUTPUT_LOW;
+			config |= PINCTRL_CONFIG_OUTPUT_HIGH;
+		}
+		config &= ~PINCTRL_CONFIG_BIAS_MASK;
+		switch (flags & GPIO_PUD_MASK) {
+		case GPIO_PUD_PULL_UP:
+			config |= PINCTRL_CONFIG_BIAS_PULL_UP;
+			break;
+		case GPIO_PUD_PULL_DOWN:
+			config |= PINCTRL_CONFIG_BIAS_PULL_DOWN;
+			break;
+		case GPIO_PUD_NORMAL:
+		default:
+			config |= PINCTRL_CONFIG_BIAS_DISABLE;
+			break;
+		}
+		config &= ~PINCTRL_CONFIG_DRIVE_MASK;
+		config &= ~PINCTRL_CONFIG_DRIVE_STRENGTH_MASK;
+		if ((flags & GPIO_DS_LOW_MASK) == GPIO_DS_DISCONNECT_LOW) {
+			config |= PINCTRL_CONFIG_DRIVE_OPEN_SOURCE;
+			/* low is disconnect, take high values */
+			if ((flags & GPIO_DS_HIGH_MASK) == GPIO_DS_ALT_HIGH) {
+				config |= PINCTRL_CONFIG_DRIVE_STRENGTH_7;
+			} else {
+				config |= PINCTRL_CONFIG_DRIVE_STRENGTH_DEFAULT;
+			}
+		} else if ((flags & GPIO_DS_HIGH_MASK)
+			== GPIO_DS_DISCONNECT_HIGH) {
+			config |= PINCTRL_CONFIG_DRIVE_OPEN_DRAIN;
+			/* high is disconnect, take low values */
+			if ((flags & GPIO_DS_LOW_MASK) == GPIO_DS_ALT_LOW) {
+				config |= PINCTRL_CONFIG_DRIVE_STRENGTH_7;
+			} else {
+				config |= PINCTRL_CONFIG_DRIVE_STRENGTH_DEFAULT;
+			}
+		} else {
+			config |= PINCTRL_CONFIG_DRIVE_PUSH_PULL;
+			if (((flags & GPIO_DS_LOW_MASK) == GPIO_DS_ALT_LOW)
+			|| ((flags & GPIO_DS_HIGH_MASK) == GPIO_DS_ALT_HIGH)) {
+				config |= PINCTRL_CONFIG_DRIVE_STRENGTH_7;
+			} else {
+				config |= PINCTRL_CONFIG_DRIVE_STRENGTH_DEFAULT;
+			}
+		}
+		/* Assure pin is on input before doing configuration
+		 * to prevent damage */
+		err = pinctrl_mux_set(data->pinctrl, pinctrl_pin,
```

Comment:
I'm not sure about this... yet I don't know how it should be handled !
But if, let's says the bootloader has already configured the pin and the config is vital to the system, doing this will maybe cause unwanted side effects !
Changing the config won't harm if already in output, meaning it already has been configured somewhere.

Reply:
You are right, will rethink.

Reply:
@superna9999 

> Changing the config won't harm if already in output, meaning it already has been configured somewhere.

Reverted to not switch to input before configuration.